### PR TITLE
Plugins: Store public key for plugin manifests in the database.

### PR DESCRIPTION
**Pull request by bergquist**
Storing the public key in the database instead of hardcoded will allow users to add new keys without recompiling/upgrading Grafana. 

Once this is merged we can think about download new keys/revoking keys from grafana.com / filesystem. 

ref https://github.com/grafana/grafana/issues/23619

### CODE REVIEW

File: pkg/plugins/manifest.go [hunk 1]
```diff
@@ -51,9 +27,61 @@ type pluginManifest struct {
...
+	lock       sync.Mutex
+	publicKeys map[string]ManifestKeys
+}
+
+func newManifestVerifier(sqlstore *sqlstore.SqlStore) *manifestVerifier {
+	return &manifestVerifier{sqlstore: sqlstore}
+}
+
+// getPublicKey loads public keys from the database.
+// Soon we can updated keys from https://grafana.com/api/plugins/ci/keys
```
Comment (aknuds1) [inline pkg/plugins/manifest.go:52]:
```suggestion
// Soon we can update keys from https://grafana.com/api/plugins/ci/keys
```

File: pkg/plugins/manifest.go [hunk 2]
```diff
@@ -51,9 +27,61 @@ type pluginManifest struct {
...
+			pmv.publicKeys[k.KeyID] = k
+		}
+	}
+
+	key, exist := pmv.publicKeys[keyID]
+	if exist {
+		return key.PublicKey, nil
+	}
+
+	return "", errors.New("Could not find public Key")
```
Comment (aknuds1) [inline pkg/plugins/manifest.go:79]: [Error strings should not be capitalized](https://github.com/golang/go/wiki/CodeReviewComments#error-strings):
```suggestion
	return "", errors.New("could not find public key")
```

File: pkg/plugins/manifest.go [hunk 3]
```diff
@@ -51,9 +27,61 @@ type pluginManifest struct {
...
+}
+
 // readPluginManifest attempts to read and verify the plugin manifest
 // if any error occurs or the manifest is not valid, this will return an error
-func readPluginManifest(body []byte) (*pluginManifest, error) {
+func (pmv *manifestVerifier) readPluginManifest(body []byte) (*pluginManifest, error) {
 	block, _ := clearsign.Decode(body)
```
Comment (aknuds1) [inline pkg/plugins/manifest.go:85]: Not related to this PR I know, but figure we might as well fix it :)
```suggestion
	block, err := clearsign.Decode(body)
	if err != nil {
	        return nil, err
	}
```

File: pkg/plugins/manifest.go [hunk 4]
```diff
@@ -66,30 +94,34 @@ func readPluginManifest(body []byte) (*pluginManifest, error) {
...
-		block.ArmoredSignature.Body); err != nil {
+	_, err = openpgp.CheckDetachedSignature(keyring, bytes.NewBuffer(block.Bytes), block.ArmoredSignature.Body)
+	if err != nil {
 		return nil, errutil.Wrap("failed to check signature", err)
 	}
 
 	return manifest, nil
 }
 
 // GetPluginSignatureState returns the signature state for a plugin
```
Comment (aknuds1) [inline pkg/plugins/manifest.go:115]:
```suggestion
// verifyPluginSignature verifies a plugin's signature.
```

File: pkg/plugins/manifest_test.go [hunk 5]
```diff
@@ -3,10 +3,43 @@ package plugins
...
+	testKeys := []ManifestKeys{
+		{KeyID: "keyid1", PublicKey: "publicKey1", UpdatedAt: nowEpoch, Since: nowEpoch},
+		{KeyID: "keyid2", PublicKey: "publicKey2", UpdatedAt: nowEpoch, Since: nowEpoch},
+	}
+
+	t.Run("can write keys to database", func(t *testing.T) {
+		session := mv.sqlstore.NewSession()
+		defer session.Close()
+		affected, err := session.Insert(testKeys)
+		assert.Nil(t, err)
```
Comment (aknuds1) [inline pkg/plugins/manifest_test.go:27]: It's best to halt the test if an error occurred.
```suggestion
		require.NoError(t, err)
```

File: pkg/plugins/manifest_test.go [hunk 6]
```diff
@@ -3,10 +3,43 @@ package plugins
...
+		assert.Equal(t, affected, int64(len(testKeys)))
```
Comment (aknuds1) [inline pkg/plugins/manifest_test.go:28]: The expectation comes before what you're checking in Equal:
```suggestion
		assert.Equal(t, int64(len(testKeys)), affected)
```

File: pkg/plugins/manifest_test.go [hunk 7]
```diff
@@ -3,10 +3,43 @@ package plugins
...
+	})
+
+	t.Run("can query keys from database", func(t *testing.T) {
+		keys, err := mv.getPublicKey("keyid1")
+		assert.Nil(t, err)
```
Comment (aknuds1) [inline pkg/plugins/manifest_test.go:33]:
```suggestion
		require.NoError(t, err)
```

File: pkg/plugins/manifest_test.go [hunk 8]
```diff
@@ -3,10 +3,43 @@ package plugins
...
+	t.Run("can write keys to database", func(t *testing.T) {
+		session := mv.sqlstore.NewSession()
+		defer session.Close()
+		affected, err := session.Insert(testKeys)
+		assert.Nil(t, err)
+		assert.Equal(t, affected, int64(len(testKeys)))
+	})
+
+	t.Run("can query keys from database", func(t *testing.T) {
+		keys, err := mv.getPublicKey("keyid1")
```
Comment (aknuds1) [inline pkg/plugins/manifest_test.go:32]:
```suggestion
		key, err := mv.getPublicKey("keyid1")
```

File: pkg/plugins/manifest_test.go [hunk 9]
```diff
@@ -3,10 +3,43 @@ package plugins
...
+		assert.Nil(t, err)
+		assert.Equal(t, keys, testKeys[0].PublicKey)
```
Comment (aknuds1) [inline pkg/plugins/manifest_test.go:34]:
```suggestion
		assert.Equal(t, testKeys[0].PublicKey, key)
```

File: pkg/plugins/manifest_test.go [hunk 10]
```diff
@@ -3,10 +3,43 @@ package plugins
...
+	})
+
+	t.Run("should return error if key doesn't exist", func(t *testing.T) {
+		_, err := mv.getPublicKey("missing keyId")
+		assert.NotNil(t, err)
```
Comment (aknuds1) [inline pkg/plugins/manifest_test.go:39]:
```suggestion
		assert.Error(t, err)
```

File: pkg/services/sqlstore/migrations/plugin_setting.go [hunk 11]
```diff
@@ -38,4 +43,54 @@ func addAppSettingsMigration(mg *Migrator) {
...
+fk0ManCIZuc0jAIJAVw2xdLr4ZQqPUhubrUyFcqlWoW8dQoQagwO8s8ubmby
+KuLA9FWJkfuuRQr+O9gHkDVCez3aism7zmJBqIOi38aNAgjJ3bo6leSS2jR/
+x5NqiKVi83tiXDPncDQYPymOnMhW0l7CVA7wj75HrFvvlRI/4MArlbsZ2tBn
+N1c5v9v/4h6qeA==
+=DNbR
+-----END PGP PUBLIC KEY BLOCK-----
+`
+
+	mg.AddMigration(
+		"insert public key into manifest)keys table",
```
Comment (aknuds1) [inline pkg/services/sqlstore/migrations/plugin_setting.go:90]:
```suggestion
		"insert public key into manifest_keys table",
```

**Review by aknuds1 — CHANGES_REQUESTED** [review]

File: pkg/plugins/manifest.go (hunk 3 above)
Comment (bergquist) [reply pkg/plugins/manifest.go:85]: This one tripped me as well. But the second return is not an error. 
https://godoc.org/golang.org/x/crypto/openpgp/clearsign#Decode

Comment (aknuds1) [reply pkg/plugins/manifest.go:85]: @bergquist Oops, my bad.

**Review by aknuds1 — APPROVED** [review]
LGTM

File: pkg/plugins/manifest.go (hunk 3 above)
Comment (bergquist) [reply pkg/plugins/manifest.go:85]: it's a veeeeery reasonable comment :) it's the library that's kind of weird.

**Pull request closed without merging**